runtime/j9vm/j8vmi.c

```diff
@@ -99,3 +99,114 @@ JVM_GetTemporaryDirectory(JNIEnv *env)
+
+/**
+ * Copies memory from one place to another, endian flipping the data.
+ *
+ * @param [in] env Pointer to JNI environment
+ * @param [in] srcObj Source primitive array (NULL means srcOffset represents native memory)
+ * @param [in] srcOffset Offset in source array / address in native memory
+ * @param [in] dstObj Destination primitive array (NULL means dstOffset represents native memory)
+ * @param [in] dstOffset Offset in destination array / address in native memory
+ * @param [in] size Number of bytes to copy
+ * @param [in] elemSize Size of elements to copy and flip
+ *
+ * elemSize = 1 means a straight byte copy
+ * elemSize = 2 means byte order 1,2 becomes 2,1
+ * elemSize = 4 means byte order 1,2,3,4 becomes 4,3,2,1
+ * elemSize = 8 means byte order 1,2,3,4,5,6,7,8 becomes 8,7,6,5,4,3,2,1
+ */
+void JNICALL
+JVM_CopySwapMemory(JNIEnv *env, jobject srcObj, jlong srcOffset, jobject dstObj, jlong dstOffset, jlong size, jlong elemSize)
+{
+	void *srcBytes = NULL;
+	void *dstBytes = NULL;
+	U_8 *srcAddr = NULL;
+	U_8 *dstAddr = NULL;
+	/* Ensure that size is a non-negative multiple of elemSize */
+	Assert_SC_true(size >= 0);
+	Assert_SC_true(0 == (size & elemSize));
+	if (NULL != srcObj) {
+		/* Disallow negative object offsets */
+		Assert_SC_true(srcOffset >= 0);
+		srcBytes = (*env)->GetPrimitiveArrayCritical(env, srcObj, NULL);
+		srcAddr = srcBytes;
+	}
+	if (NULL != dstObj) {
+		/* Disallow negative object offsets */
+		Assert_SC_true(dstOffset >= 0);
+		dstBytes = (*env)->GetPrimitiveArrayCritical(env, dstObj, NULL);
+		dstAddr = dstBytes;
+	}
+	srcAddr += (UDATA)srcOffset;
+	dstAddr += (UDATA)dstOffset;
+/*
+	Generic code would be:
+
+		jlong elemCount = size / elemSize;
+		for (jlong i = 0; i < elemCount; ++i) {
+			for (UDATA j = 0; j < elemSize; ++j) {
+				dstAddr[j] = srcAddr[elemSize - j - 1];
+			}
+			srcAddr += elemSize;
+			dstAddr += elemSize;
+		}
+		break;
+	}
+ */
+	switch(elemSize) {
+	case 1:
+		memcpy(dstAddr, srcAddr, (size_t)size);
+		break;
+	case 2: {
+		jlong elemCount = size / 2;
+		while (0 != elemCount) {
+			dstAddr[0] = srcAddr[1];
+			dstAddr[1] = srcAddr[0];
+			srcAddr += 2;
+			dstAddr += 2;
+			elemCount -= 1;
+		}
+		break;
+	}
+	case 4: {
+		jlong elemCount = size / 4;
+		while (0 != elemCount) {
+			dstAddr[0] = srcAddr[3];
+			dstAddr[1] = srcAddr[2];
+			dstAddr[2] = srcAddr[1];
+			dstAddr[3] = srcAddr[0];
+			srcAddr += 4;
+			dstAddr += 4;
+			elemCount -= 1;
+		}
+		break;
+	}
+	case 8: {
+		jlong elemCount = size / 8;
+		while (0 != elemCount) {
+			dstAddr[0] = srcAddr[7];
+			dstAddr[1] = srcAddr[6];
+			dstAddr[2] = srcAddr[5];
+			dstAddr[3] = srcAddr[4];
+			dstAddr[4] = srcAddr[3];
+			dstAddr[5] = srcAddr[2];
+			dstAddr[6] = srcAddr[1];
+			dstAddr[7] = srcAddr[0];
+			srcAddr += 8;
+			dstAddr += 8;
+			elemCount -= 1;
+		}
+		break;
+	}
+	default:
+		Assert_SC_unreachable();
+		break;
+	}
+	if (NULL != srcObj) {
+		(*env)->ReleasePrimitiveArrayCritical(env, srcObj, srcBytes, 0);
```

Comment:
Shouldn't the final argument to `ReleasePrimitiveArrayCritical()` be `JNI_COMMIT` here and in the similar call below?

Reply:
No. ABORT means do nothing, COMMIT means copy back but don't free the back buffer (if it was allocated). 0 (no special flags) is the right answer virtually 100% of the time.

Reply:
Sorry, you are in fact correct. Since this is the source array, there's no need to copy back to the heap. I'm not sure it matters in this case, since we release the destination array afterwards, so if the source and dest were the same array, only one copy back would take place. I've update this in a later PR.

Reply:
To make sure we're on the same page, the source array should be released with `JNI_ABORT` and the destination with mode `0`, right?

Reply:
Right - ABORT means free the buffer but don't copy back to heap (assuming that we didnt get dirtect access to the heap). Since we're not modifying the source array unless it's identical to the dest array, there's no need to copy back.